### DISCUSSION TITLE
Move platform domain constants to const.py

### DISCUSSION THREAD

**Discussion by cdce8p** [home-assistant/architecture#663] [General] [answered]
I've recently noticed that a lot of integrations import multiple different domains for entry setup and use in `hass.config_entries.async_setup_platforms(entry, PLATFORMS)`. Oftentimes this is overly verbose, here from [wled](https://github.com/home-assistant/core/blob/00b1c2bb70a781241f8a68b9429e4e1dccf0c379/homeassistant/components/wled/__init__.py#L4-L9).

```py
from homeassistant.components.button import DOMAIN as BUTTON_DOMAIN
from homeassistant.components.light import DOMAIN as LIGHT_DOMAIN
from homeassistant.components.number import DOMAIN as NUMBER_DOMAIN
from homeassistant.components.select import DOMAIN as SELECT_DOMAIN
from homeassistant.components.sensor import DOMAIN as SENSOR_DOMAIN
from homeassistant.components.switch import DOMAIN as SWITCH_DOMAIN

PLATFORMS = (
    BUTTON_DOMAIN,
    LIGHT_DOMAIN,
    SELECT_DOMAIN,
    SENSOR_DOMAIN,
    SWITCH_DOMAIN,
    NUMBER_DOMAIN,
)
```

Not only does each domain need to be imported from a different package, it then also needs to be reassigned to avoid name conflicts.

Another result, many don't even bother and use strings instead. (Might also be that this is the legacy version and some haven't been updated yet.) Here from [smarttub](https://github.com/home-assistant/core/blob/00b1c2bb70a781241f8a68b9429e4e1dccf0c379/homeassistant/components/smarttub/__init__.py#L5).

```py
PLATFORMS = ["binary_sensor", "climate", "light", "sensor", "switch"]
```

--
I propose we move all **platform** domains (`light`, `sensor`, `switch`, ...) to `const.py` as `DOMAIN_<...>`. This will make it much easier to import and work with them. Domains for individual integrations would **not** be moved. For backwards compatibility, domains will be reexported in the integration init files:
```py
# homeassistant/components/sensor/__init__.py
from homeassistant.const import DOMAIN_SENSOR as DOMAIN
```
Additional steps would include
* Updating existing integrations / import statements
* Replace existing string domains (in `PLATFORMS`) with constants
* Update dev docs

--
Draft PR illustrating the necessary changes: https://github.com/home-assistant/core/pull/59561

**Comment by balloob** [comment]
Current rule is that all integrations are required to specify `DOMAIN` in their `__init__.py`. It's a rule from before we had manifests.

Btw, I am totally fine with using strings in `PLATFORMS`.

**Reply by cdce8p** [reply]
> Current rule is that all integrations are required to specify `DOMAIN` in their `__init__.py`. It's a rule from before we had manifests.

In general, I think that still makes sense. I'm only suggesting to move these 22 [here](https://github.com/home-assistant/core/blob/20dd01ea96844de15465d52c9fc12ae684c7693b/homeassistant/const.py#L43-L64).

**Reply by epenet** [reply]
@balloob when you say that you are fine using strings, is it a "recommendation" that you prefer strings (in which case I should stop recommending constants in my reviews)? or is it that you do not mind either way?

**Comment by epenet** [comment]
I have been recommending constants for the PLATFORMS rather than strings in my review comments, to make it more explicit, so +1 on the idea.

Part of me thinks that platforms should be distinguished from regular components.

Maybe use a different prefix?
```python
from homeassistant.const import (  # noqa: F401  # pylint: disable=unused-import
    PLATFORM_BUTTON as DOMAIN,
)
```

or an enum
```python
from homeassistant.const import (  # noqa: F401  # pylint: disable=unused-import
    PLATFORM.BUTTON as DOMAIN,
)
```

**Reply by scop** [reply]
enum++ in principle, however if we'd be leaving component domains as strings and not make them part of the enum too, that'd pretty much nullify the benefits of the enum.

**Reply by cdce8p** [reply]
I'm not sure using Enums would be better. Especially since "normal" integration domains should not be moved.

My intend was mostly to simplify / unify the import statements.

**Reply by scop** [reply]
Yes, as said, if the component domains are not going to move, a "partial" enum won't be of much use.

**Comment by epenet** [comment]
I think that in my first answer the example were not very good, and when I worked on new example I realised that creating 22 individual constants was not very useful, and still required 6 imports to create the tuple of 6 items.

How about something using an enum like this, which makes it easier for developers to find out which platforms are considered "standard"?
```python
from homeassistant.const import PLATFORM

PLATFORMS = (
    PLATFORM.BUTTON,
    PLATFORM.LIGHT,
    PLATFORM.NUMBER,
    PLATFORM.SELECT,
    PLATFORM.SENSOR,
    PLATFORM.SWITCH,
)
```

**Reply by epenet** [reply]
Note: this generates a listing of 47 components using the relative imports in various forms (and that's only those which are inline):
`grep "PLATFORMS = [(\[][^\"]" homeassistant/components/**/*.py`

**Reply by epenet** [reply]
Note 2: this generates a list of 11 more components using the relative imports in various forms (not inline):
`grep -Pzo "PLATFORMS = [(\[]\n\s\s\s\s[^\"]" homeassistant/components/**/*.py`

**Comment by frenck** [comment, marked as answer]
Implementation suggestion for this (thank to @epenet on Discord for pushing in this direction, so credit are for him)

https://github.com/home-assistant/core/pull/60818

![image](https://user-images.githubusercontent.com/195327/144382429-ba23286b-3e25-4fa4-87d5-c6bffb9ddaa2.png)
